### ISSUE TITLE
Abrupt chapter transitions in Thorium

### ISSUE THREAD

**Issue by RobJacobson** [readium/readium-desktop#901]
I'm submitting a combination bug report/question.

The Thorium desktop reader on Windows 10 has unexpected behavior (in my opinion) during chapter transitions.  If you flip through a book with many chapters, especially using a two-page spread, you'll notice the following:

1. Each chapter begins on the left page of the spread.  At the end of the chapter, there's often a blank right-hand side column, so the next chapter begins on the next left-hand page.

2. When you reach an end-of-the-chapter spread that has a blank right column, the animation for the page transition doesn't look right.  The ending text in the chapter seems to emerge from the center of the spread and move into the left column.

3. When you flip the spread to the next chapter, there isn't a page-turn animation.  The new chapter just shows up immediately.

It seems that Thorium is simply taking each chapter as a separate HTML document and displaying them one at a time.  It doesn't cache the next chapter to be read inline with the current chapter.  This makes the reading experience seem "weird," in my opinion.

Are there any workarounds in JavaScript for these behaviors?  In particular, is there some preferred way to cache the next chapter so that it is shown on the same spread as the previous chapter?  In other words, to display the start of the next chapter in the right-hand column instead of showing an empty column?

One option would simply be to concatenate the chapters together in the DOM to form a single HTML page that represents the entire book.  That seems like a bad idea for memory and performance, though.  It would also upset the authors' intentions if (for some reason) they have different CSS styles in different chapters, since only the CSS from one chapter would be used.

Thanks very much for any insights.

**Comment by danielweck** [comment]
Hello again @RobJacobson ! :)

> Each chapter begins on the left page of the spread. At the end of the chapter, there's often a blank right-hand side column, so the next chapter begins on the next left-hand page.

When a reflowable HTML document is laid out with two-page spreads, an odd number of synthetic pages ("columns") logically results in an empty right-hand-side page slot inside the last spread (in other words, the last synthetic page of the HTML document is displayed in the left-hand-side slot of the spread ... the rest is just blanking space).
The rendering engine makes no attempt to collate adjacent "chapters" / EPUB spine items side-by-side in a seamless manner. This Readium implementation currently does not support laying out multiple webviews (i.e. containers for contiguous HTML documents) inside the viewport, so only a single chapter is displayed at any given time.

**Comment by danielweck** [comment]
> When you reach an end-of-the-chapter spread that has a blank right column, the animation for the page transition doesn't look right. The ending text in the chapter seems to emerge from the center of the spread and move into the left column.

Indeed. Due to technical limitations with the way CSS Columns are laid out in the HTML body, in cases where a paginated reflowable document results in an odd number of synthetic pages, and the document is rendered as two-page spreads, a CSS trick is used to shift the last column into its dedicated slot (i.e. the left-hand side of the spread). The right-hand-side empty space is in fact not a part of the currently-rendered document, it is literally space-filler (thus why the tail end of the HTML document appears to scroll into view from the middle of the page spread).

**Comment by danielweck** [comment]
> When you flip the spread to the next chapter, there isn't a page-turn animation. The new chapter just shows up immediately.

Yes indeed, inter-"chapter" transitions (i.e. seamless visual effect between adjacent EPUB spine items) is not supported due to displaying only a single webview container in the viewport. Conversely, intra-"chapter" transitions (i.e. horizontal scroll animation between synthetic pages / CSS columns) occur within a given webview container.

**Comment by danielweck** [comment]
> One option would simply be to concatenate the chapters together in the DOM to form a single HTML page that represents the entire book. That seems like a bad idea for memory and performance, though. It would also upset the authors' intentions if (for some reason) they have different CSS styles in different chapters, since only the CSS from one chapter would be used.

Indeed :)

The "safe" method to create the visual effect of instant inter-"chapter" transition is to prefetch and pre-render adjacent pre- and post- HTML documents in their own respective webviews, using a pool of reusable webviews in order to mitigate the inevitable performance degradation. There are also issues with CSS animations and/or Javascript behaviours that should only be started when documents come into view (i.e. not whilst they are loaded "in the background" / out of view). Note that even if this Readium implementation continues to limit a single webview per viewport (i.e. no seamless adjacent EPUB spine items), using pre-loaded webviews instead of loading a single one on-demand would yield significant "perceived performance" improvements (i.e. smoother, more pleasant user experience).

So, this feature was considered in previous development iterations, but there were technical challenges at the time (i.e. older version of Electron / Chromium), and this was considered a "nice to have" feature compared with higher-priority todo list items. I will try to revisit this problem space later next year, but we already have a pretty crowded development plan in Q1, so I would rather not give an ETA. Note that this work might overlap with the planned improvements of support for fixed-layout EPUBs (this Readium implementation currently only supports single-page FXL rendering)

**Comment by RobJacobson** [comment]
Thank you so much for the detailed explanations!  I'm glad this is on your radar.  I understand the technical rational (CSS multicol has limitations).  But for an optimal book-like reading experience, it seems desirable to have the option of showing facing pages from two chapters on a spread.

For what it's worth, I noticed that the Vivliostyle Viewer allows this feature.  However, I haven't studied their rendering engine to know how it works (whether CSS multicol or something else).

Thanks for maintaining this great project!

**Comment by RobJacobson** [comment]
Could I revisit the issue of having a two-page spread that spans two chapters?  I've been thinking about this problem space and doing some research.  I could perhaps work on the issue and submit a pull request, but I'm not sure even where to start.

Both Vivliostyle Viewer and Kindle Cloud Reader can show two chapters on facing pages.  But when I looked into their implementations, they're not using CSS multicol for their layouts.  Instead, they seem to do layout the "hard way" by parsing the HTML into spans of individual words and then determining how many spans can fit on a single page.

To get this working with CSS multicol, the chief problem seems to be determining either "how many column boxes does this chapter require?" or, equivalently, "does the content end on a left page or a right page?"  If a chapter starts on the left-hand side and has an odd number of "pages," then the chapter ends on a left-hand page.  Thus, there would be space to begin a new chapter on the next right-hand page.  You could potentially use a pool of just two WebViews, one to show the current chapter and the other for the next chapter.

However, I can't find any way to count the number of required "pages" with CSS/JavaScript.  According to [this article](https://www.smashingmagazine.com/2019/01/css-multiple-column-layout-multicol/), CSS multicol does not expose many properties to JavaScript, and I couldn't find any that would support this calculation.

Would you have any general ideas about how to approach this problem?  If you could point me in the right direction, I could try to run with it.

Thanks again!

**Comment by RobJacobson** [comment]
One other thought:

It might be worth exploring and testing whether it's practical to combine all of the chapters/spine items into a single HTML page in the DOM:

1. Memory and performance:  This approach could conceivably degrade performance for the browser/WebView.  Since most epub books are fairly small, though, this might be more of a theoretical problem (especially on desktops or notebooks with ample CPU and memory resources).  There might be edge cases for books that have high-resolution photos or embedded videos, but those seem to be uncommon.  It might be interesting to compare the performance of the current implementation (i.e., many spine items) versus this proposal (a single, monolithic spine item that gets rendered).

2.  Different CSS styles for different chapters:  If different chapters have different CSS styles, we could develop a parser or preprocessor that detects these differences and creates unique styles for each spine item.  E.g., if there are different styles for <p> between Chapter 1 and Chapter 2, we could programmatically create a single stylesheet with separate classes for "Chapter1 p" and "Chapter2 p."  This would also conceivably degrade performance, but it might be worth testing to see if this degredation is real or just hypothetical in the typical case.

3.  Concerning "CSS animations and/or Javascript behaviours that should only be started when documents come into view":  That's a valid concern, but it might also be an edge case.  This might also be a concern with the current implementation, insofar as an element on, e.g., page 19 of a particular chapter won't come into view immediately, but only after the user reads pages 1-18.  Does Readium currently have a mitigation against this issue within individual chapters?

Thanks again!

**Comment by RobJacobson** [comment]
There doesn't seem to be any interest here, so I'll close the issue.  I hope you can get this working.
